### PR TITLE
Moving log4j to logback [1/n]

### PR DESCRIPTION
**Pull request by rahulrane50**
### Description
Currently zk application and audit logs are rotated based on size and not on time/duration. For GCN debugging sometimes it's hard to debug things when logs are rolled over and cleaned up because current policy. Currently log4j do not allow to have rollover policy for size as well as duration based rotation. Even upstream community is moved to logback after discussing pros and cons about log4j 2.x and logback https://github.com/apache/zookeeper/pull/1793.

This PR is based out of PR https://github.com/apache/zookeeper/pull/1793. We havn't pulled latest changes from upstream zookeeper so i reverted few changes and kept it compatible with our current zk implementation.

### Tests
CI and e2e pipeline

Verified locally running zk server and connecting via some internal tools and performing crwd ops.
<using shell to connect to local zk server> [INFO] Hosts: ['127.0.0.1:2181'] (CONNECTED [127.0.0.1:2181]) /> ls TestCluster ephe1 folder nomapping test3 test4 test5 test6 test7 testwithacl testwithacl1 zookeeper (CONNECTED [127.0.0.1:2181]) /> create test1 lskd (CONNECTED [127.0.0.1:2181]) /> get test1 lskd (CONNECTED [127.0.0.1:2181]) /> set test1 ksdj (CONNECTED [127.0.0.1:2181]) /> rm test1 (CONNECTED [127.0.0.1:2181]) />

### Changes that Break Backward Compatibility (Optional)
NA

### Documentation (Optional)
In case of new functionality, my PR adds documentation in the following wiki page:
(Link the GitHub wiki you added)

### CODE REVIEW

**Comment by rahulrane50** [comment]
Just for the record there is only flaky C test and it's been fixed in master https://issues.apache.org/jira/browse/ZOOKEEPER-4479. We have to cherry pick this commit to prevent our efforts of rerunning CI.

File: conf/logback.xml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    <encoder>
+      <pattern>%d{ISO8601} [myid:%X{myid}] - %-5p [%t:%C{1}@%L] - %m%n</pattern>
+    </encoder>
+    <filter class="ch.qos.logback.classic.filter.ThresholdFilter">
+      <level>TRACE</level>
+    </filter>
+  </appender-->
+
+  <!--
+    zk audit logging
```
Comment (mgao0) [inline conf/logback.xml:84]: Do we need this?

File: zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java [hunk 2]
```diff
@@ -322,8 +312,7 @@ private static void verifyLog(String expectedLog, String log) {
         String searchString = " - ";
         int logStartIndex = log.indexOf(searchString);
         String auditLog = log.substring(logStartIndex + searchString.length());
-        assertEquals(expectedLog, auditLog);
-
+        Assert.assertTrue(auditLog.endsWith(expectedLog));
```
Comment (mgao0) [inline zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java:315]: Just curious, why with log4j it's "equals", and with logback it's "endWith"? Does logback add something in the front?

File: zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java [hunk 3]
```diff
@@ -431,12 +420,5 @@ public static void tearDownAfterClass() {
                 e.printStackTrace();
             }
         }
-        try {
-            zlogger.removeAppender(appender);
-            os.close();
```
Comment (mgao0) [inline zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java:436]: Do we need to close os?

File: conf/logback.xml (hunk 1 above)
Comment (rahulrane50) [reply conf/logback.xml:84]: Yup we need this. For now i kept it in comments and next few PRs should have correct audit and console logging enabled.

File: zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java (hunk 2 above)
Comment (rahulrane50) [reply zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java:315]: yeah logback logging format is little different hence we had to check with endsWith. I verified logback had all existing things from log4j in logs but just few extra things like classname in detail or few formatting changes.

File: zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java (hunk 3 above)
Comment (rahulrane50) [reply zookeeper-server/src/test/java/org/apache/zookeeper/audit/Slf4JAuditLoggerTest.java:436]: good catch!

File: bin/zkEnv.cmd [hunk 4]
```diff
@@ -16,7 +16,6 @@ REM limitations under the License.
 
 set ZOOCFGDIR=%~dp0%..\conf
 set ZOO_LOG_DIR=%~dp0%..\logs
-set ZOO_LOG4J_PROP=INFO,CONSOLE
```
Comment (desaikomal) [inline bin/zkEnv.cmd:19]: is there equivalent to set log-levels to o/p mapping in logback?

File: zookeeper-server/src/test/java/org/apache/zookeeper/test/ReadOnlyModeTest.java [hunk 5]
```diff
@@ -198,6 +205,9 @@ public void testConnectionEvents() throws Exception {
      */
     @Test(timeout = 90000)
     public void testSessionEstablishment() throws Exception {
+        qu.enableLocalSession(true);
```
Comment (desaikomal) [inline zookeeper-server/src/test/java/org/apache/zookeeper/test/ReadOnlyModeTest.java:208]: this is not related to logback change? any info as to why we had to add this?

Comment (rahulrane50) [reply zookeeper-server/src/test/java/org/apache/zookeeper/test/ReadOnlyModeTest.java:208]: Looks like few APIs have been changed due to which this code has to be added to pass this UT

File: bin/zkEnv.cmd (hunk 4 above)
Comment (rahulrane50) [reply bin/zkEnv.cmd:19]: yup that's correct. And that can be done using xml file.

**Review by mgao0 — APPROVED** [review]
